src/leetCodeSolutionProvider.ts

```diff
@@ -1,49 +1,31 @@
 // Copyright (c) jdneo. All rights reserved.
 // Licensed under the MIT license.
 
-import * as hljs from "highlight.js";
-import * as MarkdownIt from "markdown-it";
-import * as path from "path";
-import * as vscode from "vscode";
+import { TokenRender } from "markdown-it";
 import { Disposable, ExtensionContext, ViewColumn, WebviewPanel, window } from "vscode";
-import { leetCodeChannel } from "./leetCodeChannel";
 import { IProblem } from "./shared";
+import { MarkdownEngine } from "./webview/markdownEngine";
 
 class LeetCodeSolutionProvider implements Disposable {
 
     private context: ExtensionContext;
     private panel: WebviewPanel | undefined;
-    private markdown: MarkdownIt;
-    private markdownPath: string; // path of vscode built-in markdown extension
+    private markdown: MarkdownEngine;
     private solution: Solution;
 
     public initialize(context: ExtensionContext): void {
         this.context = context;
-        this.markdown = new MarkdownIt({
-            linkify: true,
-            typographer: true,
-            highlight: this.codeHighlighter.bind(this),
-        });
-        this.markdownPath = path.join(vscode.env.appRoot, "extensions", "markdown-language-features");
-
-        // Override code_block rule for highlighting in solution language
-        // tslint:disable-next-line:typedef
-        this.markdown.renderer.rules["code_block"] = (tokens, idx, options, _, self) => {
-            const highlight: string = options.highlight(tokens[idx].content, undefined);
-            return [
-                `<pre><code ${self.renderAttrs(tokens[idx])} >`,
-                highlight || this.markdown.utils.escapeHtml(tokens[idx].content),
-                "</code></pre>",
-            ].join("\n");
-        };
+        this.markdown = new MarkdownEngine();
+        this.addMdDefaultHighlight(); // use solution language if highting block lang is undefined
```

Comment:
The two methods `addMdDefaultHighlight()` and `addMdImageUrlCompletion()` is somehow confusing here. The purpose of this PR is to extract and decouple the markdown related code with the webviews, but these two methods are still making them coupling together, which is not a good design.

Looks like the method `addMdImageUrlCompletion()` can be moved into the markdown engine since it does not rely on any member in the solution provider.

For `addMdDefaultHighlight()` we rely on `this.solution.lang` so I asked a question below to see that if it is possible to remove the dependency here.

Reply:
The reason why I merely export `MarkdownEngine` class definition, rather than a global variable `export const markdownEngine = new MarkdownEngine` is that there may be some customization in each WebView, e.g. `this.solution.lang` and url completion here.

Markdown-it may provide some good mechanism for adding custom rule (maybe [Ruler](https://markdown-it.github.io/markdown-it/#Ruler)), but some custom code in each WebView may not be easily avoided.

Reply:
The workaround may be store `host` and `defaultLang` as an option in MarkdownEngine or somewhere in `MarkdownIt`, then we can move them out of Solution WebView. Do you prefer this way?